Patch for trackpad scrolling on Mac

FuzzyRhombus opened the pull request:
#### Related to https://github.com/godotengine/godot/issues/492 and https://github.com/godotengine/godot/issues/3913

This is a quick patch for scrolling with trackpad on Mac (was way too fast). It is **not intended** as a permanent solution, but it is working well enough for me now and also provides a starting point if anyone else wants to continue on these issues. =]

It does **NOT** fix the slow horizontal scrolling in the editor - the editor also seems to still scroll faster vertically than other components, but it is still inherently better with the patch.

This was tested on an older 15" MBP without retina display - **I am building/testing on the 2.1 branch**

FuzzyRhombus committed:
quick fix for trackpad scroll events on mac

FuzzyRhombus committed:
always use deltaX/deltaY for scroll wheel events (scrolling* is for px); related cleanup

FuzzyRhombus committed:
handle leftover delta when momentum scroll event ends

akien-mga commented:
Any Mac user to test this? (@BastiaanOlij maybe?)

BastiaanOlij commented:
Yeah I'll have a look over the weekend:)

BastiaanOlij commented:
Finally had a chance to try this out. It works and is an improvement on how it is before applying the patch but it is a bit choppy. I tested this on a 15" Retina MBP btw.

I'm not entirely sure if I understand the fix. If I understand it correctly it's applying handling the scroll momentum after the user lets go of the trackpad by recursively calling scrollWithMomentum until both scroll accumulators hit zero. Each iteration it pushes out scroll events to Godot. Even though this breaks it up into a nice end scroll wouldn't Godot just handle all the scroll events in one frame (which would explain the choppyness)?
 
I have to admit, I've never looked into the scroll events Mac OS X sends to the application so I may be missing the obvious here.

All in all I do believe the improvement itself is necessary, the current 2.1.2 build simply scroll to the end of my test list with even the slightest scroll gesture.

BastiaanOlij commented on platform/osx/os_osx.mm:
```diff
@@ -830,62 +832,65 @@ - (void)keyUp:(NSEvent *)event
+    bool endScroll = false;
+
+    if (hasMomentum) {
+        switch (event.phase) {
```
According to https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/EventOverview/HandlingTouchEvents/HandlingTouchEvents.html

There are two phase properties, phase (which is being checked) and momentumPhase (which should be the value checked). It's a bit confusing but phase is used for devices that do not support momentum scrolling which momentumPhase is used for devices that do support it. So the values checked here would not be used for Phase.

FuzzyRhombus replied:
It is confusing - I chose `phase` over `momentumPhase` because I would getting behavior described by those docs **but** on the `phase` property. Using the scroll wheel on the mouse gives `NSEventPhaseNone` for both properties on every event. In the case of the trackpad `momentumPhase` will give `NSEventPhaseBegan` one event later and then change `NSEventPhaseNone` for subsequent events. Odd behavior

BastiaanOlij replied:
That is how I understand it to work as well. The scrollwheel doesn't support momentum therefor both are set to NONE. When you use a Magic Mouse or Trackpad the momentum scroll is supported, phase stays NONE, momentumPhase has the values that indicate start and end, but the events in between wasn't very clear in that documentation.

It really sounds like a half arsed implementation on Apples part, though at the same time reading through its history, it sounds like they kept changing their minds on how it should work every couple of years.

BastiaanOlij commented on platform/osx/os_osx.mm:
```diff
@@ -830,62 +832,65 @@ - (void)keyUp:(NSEvent *)event
-	    deltaY *= 0.1;
-	}
+	double factor = [event hasPreciseScrollingDeltas] ? 1.0 : 0.1,
+	       deltaX = [event deltaX] * factor,
```
If I understand correctly deltaX and deltaY are both deprecated, scrollingDeltaX and scrollingDeltaY should be used.

FuzzyRhombus replied:
I can't find the link atm but I read scrollingDelta[X|Y] indicate the amount scrolled in pixels, while deltaX and deltaY are traditional line scroll amounts. Not every event will have a non-zero value for delta[X|Y] as it seems internally accumulate enough for a "line".

I should have mentioned I am on 10.10.5, and with the way Apple likes to move on you could be right

BastiaanOlij replied:
Yes indeed, deltaX/deltaY are the old variables and while deprecated will still work for awhile. They are based on the old scroll methodology where you would scroll in units (often 8 pixels).

scrollingdelta is a more precise value, its not pixels but points, important on retina displays as everything is scaled up and that may pose a problem. I haven't looked into how Godot handles retina displays, I know there are some display issues I should really start looking into caused by the scaling (buttons wrongly placed in the interface). 

But I do believe scrollingDeltaX/Y should be used going forward.

FuzzyRhombus replied:
Ah makes sense, agreed then - I suppose that is why the conditional macros were there in the first place (whoops). In theory then only the factor/threshold values need to be adjusted to use scrollingDeltaX/Y when available

BastiaanOlij replied:
Yes but at the same time those conditions are from a long long time ago, for Godot 3 we already moved the minimum OSX version above 10. I would suggest removing them in Godot 3 anyway but focussing on the newer branch using scrollingDeltaX/Y.

How the factor should be applied will be very dependent on what changes can be made further inside of the core.

FuzzyRhombus commented:
Thanks for looking into this @BastiaanOlij - and no worries I am new as well - I have minimal experience with Obj-C but it was all iOS, so I came up with it checking the docs and playing with it myself.

I agree it is odd - as you saw the system will continuously send events with the momentum phase. What I found is that not each event will have a non-zero `delta[X|Y]`. I haven't really dug further into Godot yet but it seems that scrolling is handled as fixed amount button presses rather than an event with a scroll amount. So basically what this patch is doing is trying to determine if the event sent is a momentum scroll (trackpad) and if so accumulate the amount and send a scroll event press only when that accumulator has reached a certain threshold and so on (almost like a fixed timestep) - this way it is not sending a scroll event for each native event as it was before. The recursion is used when the user stops scrolling in case there is more than one "step" left to send (as there won't be any more events)

I also did notice that changing the my scrolling speed for both mouse and trackpad don't seem to affect Godot as it seems to simple be discarding that information. If I'm correct in this I don't think there will be a perfect solution without support for scroll amounts within Godot, so this is attempt to get around it

BastiaanOlij commented:
@FuzzyRhombus ah that makes a whole lot more sense then. I'm guessing the scrolling behaviour in Godot was build/designed a long time ago. Scrollbars where controlled by the buttons at each end that would move the scroll position by fixed amounts. Moving the scroll bar with the mouse would move the scroll position in equal steps.

That would indeed make it near impossible to get momentum scrolling implemented properly. I'll see if I can dive somewhat deeper in the Godot code and see if it would be easy to add more precise scrolling. That would be a good Godot 3 feature.

FuzzyRhombus commented:
@BastiaanOlij For sure and I definitely agree with the improvements for v3 - seems like it would be too much to focus on more than a patch for older versions. I can look into the other issues you brought up with if this is something you would like to use as a 2.1/2.2 fix

BastiaanOlij commented:
@FuzzyRhombus up to @akien-mga  to make a call on I'd say. Personally I think its not worth upsetting the 2.x branch for but I'd definitely like to dive deeper into this for Godot 3. Not knowing enough yet about the internals of scrolling and how this effects the other platforms I think implementing more precise scrolling into the core and properly supporting momentum scrolling both on OSX and iOS would be a good thing to do. Might even find a way to get similar behaviour on windows and linux :)

akien-mga commented:
I'm fine with having different fixes pushed to 2.1 and master, but in such a case I'd want the master fix to be pushed before the 2.1 one - we don't want an issue to be worked around in the 2.1.x if it comes back in master, that will upset users.

FuzzyRhombus commented:
I'll work on the changes discussed and give it a try on master and come back with an update then. May have a chance to try it on a retina as well

BastiaanOlij commented:
@FuzzyRhombus I've been a bit busy privately the past week and looks like me schedule isn't freeing up but if you need anything tested on a retina let me know, I can always test stuff out.

FuzzyRhombus commented:
Same @BastiaanOlij, I will be hacking at this again today or tomorrow. Good to know you have a retina :) I'll let you know if I need anything, thanks

FuzzyRhombus commented:
Apologies for not following up on this. In the meantime @toger5 came up with a proper solution in https://github.com/godotengine/godot/pull/7864. I recommend that PR over this one. @BastiaanOlij, @akien-mga, as you may have seen, his PR addresses the issues you brought up as well as implementing an actual scrolling factor and is compatible with **master**. Please check it out and I can update this PR or close it

FuzzyRhombus commented:
Closing this as https://github.com/godotengine/godot/pull/7864 has been closed
